### PR TITLE
[Coroutines] Split buildCoroutineFrame into normalization and frame building

### PR DESCRIPTION
**Pull request by TylerNowicki**
* Split buildCoroutineFrame into code related to normalization and code related to actually building the coroutine frame.
* This will enable future specialization of buildCoroutineFrame for different ABIs while the normalization can be done by splitCoroutine prior to calling buildCoroutineFrame.

See RFC for more info: https://discourse.llvm.org/t/rfc-abi-objects-for-coroutines/81057

### CODE REVIEW

**Review by yuxuanchen1997 — APPROVED** [review]
Overall I think a neutral/positive change. One thing I am worried was that how we are going to distinguish a "normalized" vs a "unnormalized" coroutine? i.e. How do we find out that we missed a normalization here?

**Comment by TylerNowicki** [comment]
> Overall I think a neutral/positive change. One thing I am worried was that how we are going to distinguish a "normalized" vs a "unnormalized" coroutine? i.e. How do we find out that we missed a normalization here?

Normalization performs these important steps:
1) split around each suspend, this adds BBs before/after each suspend so each suspend now exists in its own block.
2) split around coro.end (similar to above)
3) break critical edges and add single-edge phis in the new blocks (also removing other single-edge phis).

Each of these things can individually be tested
A) Check that each suspend is the only inst in its BB
B) Check that coro.end is the only inst in its BB
C) Check that each edge of a multi-edge phis is preceded by single-edge phi in an immediate pred

For 1) and 2) I believe the purpose of the transform is in part for suspend crossing info's analysis so it can specifically 'mark' the suspend blocks and identify the end of the coroutine. There are some existing places within suspend crossing info that visit the CoroSuspends and CoroEnds so we could check A) and B) there.

For 3) I believe the purpose of this transform is for insertSpills to work properly. Infact there is already a check for the result of this transform!

    assert(PN->getNumIncomingValues() == 1 &&
               "unexpected number of incoming "
               "values in the PHINode");

I think to verify the result of normalization we just need to add checks A) and B) to suspend crossing info.

**Comment by TylerNowicki** [comment]
> Overall I think a neutral/positive change. One thing I am worried was that how we are going to distinguish a "normalized" vs a "unnormalized" coroutine? i.e. How do we find out that we missed a normalization here?

I posted a PR to add a couple of asserts that check for normalization. The PR includes this change (I posted it too quickly), I will rebase after this change merges.

https://github.com/llvm/llvm-project/pull/108096

File: llvm/lib/Transforms/Coroutines/CoroFrame.cpp [hunk 1]
```diff
@@ -1754,7 +1754,8 @@ static bool willLeaveFunctionImmediatelyAfter(BasicBlock *BB,
   if (depth == 0) return false;
 
   // If this is a suspend block, we're about to exit the resumption function.
-  if (isSuspendBlock(BB)) return true;
+  if (isSuspendBlock(BB))
+    return true;
```
Comment (ChuanqiXu9) [inline llvm/lib/Transforms/Coroutines/CoroFrame.cpp:1758]: nit: generally we don't unrelated changes.

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM

**Comment by ChuanqiXu9** [comment]
I think the concept `normalization` is pretty narrow in CoroFrame itself (or even buildCoroutineFrame itself). It is simply some transformations to make it easier to construct and use the frame. So I didn't take it as a formal concept.
